File: apps/android/app/src/main/java/ai/openclaw/app/gateway/GatewayCustomHeaders.kt
package ai.openclaw.app.gateway

/**
 * Operator-defined HTTP headers attached to gateway connections so gateways fronted by
 * authenticating reverse proxies (Cloudflare Access-style service tokens) stay reachable.
 * Header values are credentials: persist them only in SecurePrefs and never log them.
 */
object GatewayCustomHeaders {
  // Connection-management headers the WebSocket upgrade owns. Operator overrides here would
  // corrupt the handshake or duplicate fields OkHttp sets itself.
  private val reservedNames =
    setOf("connection", "content-length", "host", "proxy-connection", "upgrade")
  private const val RESERVED_PREFIX = "sec-websocket-"
  private const val TOKEN_PUNCTUATION = "!#$%&'*+-.^_`|~"

  fun isReservedName(name: String): Boolean {
    val normalized = name.trim().lowercase()
    return normalized in reservedNames || normalized.startsWith(RESERVED_PREFIX)
  }

  /**
   * Drops entries that cannot travel as a single well-formed header: empty, reserved, or
   * non-token names, and values outside printable ASCII. Dropping invalid entries keeps one bad
   * stored value from wedging every reconnect or being interpreted differently by a proxy.
   */
  fun sanitized(headers: Map<String, String>): Map<String, String> {
    val result = LinkedHashMap<String, String>()
    for ((rawName, value) in headers) {
      val name = rawName.trim()
      if (name.isEmpty() || isReservedName(name)) continue
      if (!name.all(::isTokenCharacter)) continue
      if (!value.all { it in ' '..'~' }) continue
      result[name] = value
    }
    return result
  }

  private fun isTokenCharacter(character: Char): Boolean =
    character in '0'..'9' ||
      character in 'A'..'Z' ||
      character in 'a'..'z' ||
      character in TOKEN_PUNCTUATION
}

File: apps/shared/OpenClawKit/Sources/OpenClawKit/GatewayCustomHeaders.swift
import Foundation

/// Operator-defined HTTP headers attached to every gateway-bound request so gateways fronted
/// by authenticating reverse proxies (Cloudflare Access-style service tokens) stay reachable.
/// Header values are credentials: persist them in the platform secure store and never log them.
public enum GatewayCustomHeaders {
    /// Connection-management headers the WebSocket upgrade owns. Operator overrides here would
    /// corrupt the handshake or smuggle conflicting transport state past URLSession.
    private static let reservedNames: Set<String> = [
        "connection", "content-length", "host", "proxy-connection", "upgrade",
    ]
    private static let reservedPrefix = "sec-websocket-"
    private static let tokenPunctuation = Set("!#$%&'*+-.^_`|~".utf8)

    public static func isReservedName(_ name: String) -> Bool {
        let normalized = name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return self.reservedNames.contains(normalized) || normalized.hasPrefix(self.reservedPrefix)
    }

    /// Drops entries that cannot travel as a single well-formed header: empty, reserved, or
    /// non-token names, and values with control characters (request-splitting guard).
    public static func sanitized(_ headers: [String: String]) -> [String: String] {
        var result: [String: String] = [:]
        for (name, value) in headers {
            let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmedName.isEmpty,
                  !self.isReservedName(trimmedName),
                  self.isValidName(trimmedName),
                  !self.containsControlCharacters(value)
            else { continue }
            result[trimmedName] = value
        }
        return result
    }

    private static func isValidName(_ name: String) -> Bool {
        name.utf8.allSatisfy { byte in
            switch byte {
            case 48...57, 65...90, 97...122:
                true
            default:
                self.tokenPunctuation.contains(byte)
            }
        }
    }

    private static func containsControlCharacters(_ text: String) -> Bool {
        text.unicodeScalars.contains { $0.properties.generalCategory == .control }
    }
}
